### ISSUE TITLE
Update `join` and `leave` subnet

### ISSUE THREAD

**Issue by cryptoAtwill** [consensus-shipyard/ipc-solidity-actors#203]
### Motivation
As we are relying on `fendermint` to perform top down and bottom up checkpointing, the existing way to update validator set and collateral management might have to be updated. 
Checkpoints are submitted at a fixed interval and validator sets are updated at the fixed interval. If the validator leaves the subnet and withdraws the collateral immediately, one cannot guarantee the validator would behave legitimately. We should not allow validator to withdraw immediately and only withdraw the collateral after the next checkpoint is committed in the child chain.

#### Parent Chain
When `join` and `leave` is called, the changes are committed to chain (we can also just emit an event, but based on the current util infra, committing to chain is faster to implement). For example, we can have `ValidatorChangeSet` struct in the subnet with something like the below:
```solidity
struct ValidatorChangeSet {
    /// WIth each update, the configuration number increments by 1
    configuration_number: u64,
    operation: Join/Leave,
    worker_address: FvmAddress,
    address: address,
    /// Some extra data payload the callers should interpret themselves
    extra_payload: bytes
}
``` 
Then we could have an array that keeps the changes:
```solidity
validatorChanges: ValidatorChangeSet[]
```
Then `join` and `leave` are just simply appending the `validatorChanges`.

#### Top Down Finality
We add a getter function `getValidatorChanges` to get the unexecuted validator changes, this will be used by top down parent finality.
We add an `applyValidatorChanges` function in the `GatewayManagerFacet` that applies the changes, like how it's done currently. In `fendermint` top down parent finality, after the proposals are finalised, it will call `applyValidatorChanges`. This relays the validator set changes to the child chain.

#### Bottom Up Finality
In bottom up checkpoint, the validators submits the checkpoint at a fixed interval. For the next submission epoch, validators need to query the validator set in the child chain from the `SubnetActor` contract. Validators will obtain the last executed configuration number for that epoch and its corresponding validator set, we name this set as `OldValidatorSet`. The signatures for the bottom up checkpoint must be collected from this group of `OldValidatorSet` and propagate to the parent chain. 
_**(Question1: what if we cannot get all the signatures from all the validators in `OldValidatorSet`? Will it block the bottom up checkpoint in anyway? Or we just need from some of the validators?)**_

We also need to add one more field to the parent gateway
```solidity
childAppliedConfigurationNumber: u64
```
Validators can only withdraw their collateral after the configuration number in their `leave` request is later earlier than the `childAppliedConfigurationNumber`.

**Comment by adlrocha** [comment]
Thank you for the write-up, @cryptoAtwill.

> We should not allow validator to withdraw immediately and only withdraw the collateral after the next checkpoint is committed in the child chain.

This means that all `leave` operations will be queued and only triggered in batch once a new checkpoint is submitted, right? I think that is the right behavior, especially once slashing becomes part of the system.

**Comment by cryptoAtwill** [comment]
@adlrocha Yeah, that's the idea. Withdraw of stake will be queued and only enough checks are performed.

**Comment by cryptoAtwill** [comment]
For question:
- What if we cannot get all the signatures from all the validators in OldValidatorSet? Will it block the bottom up checkpoint in anyway? Or we just need from some of the validators?

See https://github.com/consensus-shipyard/ipc-solidity-actors/pull/202/files#r1328622351

**Comment by cryptoAtwill** [comment]
See comment from @aakoshh in https://github.com/consensus-shipyard/ipc-solidity-actors/pull/204#pullrequestreview-1635201736

**Comment by cryptoAtwill** [comment]
Based on the above comment and [discussions](https://filecoinproject.slack.com/archives/C04JR5R1UL8/p1695287966581709), two contracts, `Staking` and `WithdrawQueue` is added to handle the validators‘ collaterals. 

At a high level:
- `Staking` manages the deposit and withdraw requests from existing validators and potential validators.
- `WithdrawLocking` locks the funds approved to withdraw in `Staking` contract for certain amount of time to enhance security.

The rough designs are:
```solidity
Library ValidatorSet {
    // Set the number of active validators allowed, i.e cometBFT supports 100+ validators
    function setActiveValidatorsSize(uint16 size);

    // Get the list of validators that are currently producing blocks
    function listActiveValidators() returns (Validator[]);

    // Get the list of validators that are waiting to be `promoted` to active, 
    // because they don't have as high collateral
    function waitingValidators() returns (Validators[]);

    function getCollateral(address validator) returns uint256;

    // Deposit the collateral of the validator.
    function deposit(address validator, uint256 collateral);

    // Withdraw the collateral from the validator.
    function withdraw(address validator, uint256 collateral);

    // Reshuffle the active validators and waiting validators. Active validators might be 
    // knocked off due to collateral withdraw while waiting validators can be promoted
    function reshuffle();
}

contract WithdrawLocking {
    uint256 lockingBlocks;

    // Set the amount and time for release collateral
    function setReleaseTime(address validator, uint256 amount);

    // Validator claim the collateral that are released
    function claim(uint256 amount);

    // Potentially slashing function
    function slash(address validator, uint256 amount);
}

contract Staking is WithdrawLocking {
    ValidatorSet validators;
    // Contains the list of changes to validator set. Configuration number
    // is associated at each change.
    ValidatorChangeSet changeSet;

    // Deposit `msg.value` by `msg.sender` into the collateral
    function deposit() {
        // we don't allow withdraw after deposit requests
        if (changeSet.hasWithdrawRequest(msg.sender)) {
            revert;
        }

        // each insert will increment the configuration number by 1, update will not.
        changeSet.upsertDeposit(msg.sender, msg.value);
        transferIn();
    }

    // Withdraw `msg.value` amount of collateral by `msg.sender`
    function withdraw() {
        // we don't allow deposit after withdraw requests
        if (changeSet.hasDepositRequest(msg.sender)) {
            revert;
        }

        if (validators.getCollateral(msg.sender) < msg.value) {
            revert;
        }

        // each insert will increment the configuration number by 1, update will not.
        changeSet.upsertWithdraw(msg.sender, msg.value);
    }

    // Confirm the changes in bottom up checkpoint submission, 
    // only call this in bottom up checkpoint execution.
    function confirmChange() internal {
        for (operation, validator, amount) in changeSet {
            if (operation == Withdraw) {
                validators.withdraw(validator, amount);
                setReleaseTime(validator, amount);
            } else {
                validators.deposit(validator, amount); 
           }
        }
        
        validators.reshuffle();
    }
}
```
